### PR TITLE
Graceful Logout

### PR DESCRIPTION
**Pull request by RobertDenis**
Adding graceful logout and dialog.

### CODE REVIEW

**Comment by ahocevar** [comment]
Currently the countdown dialog is shown right after launching Composer. Was this intentional?

**Comment by eshon** [comment]
Also are GeoServer session timeouts reset after each call? If so they need to be updated from backend.js in the client.

**Comment by eshon** [comment]
I don't get the modal on login now, but when it appears and I select "Stay logged in", is reappears several times and I have to reselect "Stay logged in" several times.

**Comment by eshon** [comment]
@ischneider do you have time to review this PR?

**Comment by RobertDenis** [comment]
Jira #53.

File: geoserver/webapp/app/components/alertpanel/alertpanel.js [hunk 1]
```diff
@@ -43,24 +44,26 @@ angular.module('gsApp.alertpanel', [
...
+                size: 'lg',
+                resolve: {
+                  message: function() {
+                    return message;
+                  }
+                },
+                controller: function($scope, $modalInstance, message) {
+                  $scope.message = message;
+                  $scope.copy = function(message) {
+                    $modalInstance.close();
```
Comment (ischneider) [inline geoserver/webapp/app/components/alertpanel/alertpanel.js]: The copy function closes?

File: geoserver/webapp/app/app.js [hunk 2]
```diff
@@ -41,14 +44,88 @@ angular.module('gsApp', [
         AppSession.clear();
         $state.go('login');
       });
+      $scope.$on(AppEvent.Timeout, function(e){
```
Comment (ischneider) [inline geoserver/webapp/app/app.js]: I don't see any `AppEvent.Timeout` in the codebase. I do see a `SessionTimeout` below but no corresponding entry in `core/events.js`

File: geoserver/webapp/app/app.js [hunk 3]
```diff
@@ -41,14 +44,88 @@ angular.module('gsApp', [
...
+
+      $scope.setSession = function() {
+        //Cancel any previous timeouts and set up a new one.
+        if ($rootScope.timeout) { $timeout.cancel($rootScope.timeout); }
+
+        if($location.path() != '/login') {
+          AppSession.update(AppSession.id, AppSession.user);
+          $rootScope.timeout = $timeout(function(){},
+            $rootScope.sessionWarning).then(function() {
+            $scope.$broadcast(AppEvent.SessionTimeout);
```
Comment (ischneider) [inline geoserver/webapp/app/app.js]: Sometimes you're using `$scope`, sometimes `$rootScope`.

If the scope is a child of the rootScope and listeners are registered on the rootScope, events won't get there.

**Comment by ischneider** [comment]
I'm not very familiar with the codebase but poked around a little bit. Besides the specific code comments, I'd want to see:

1. app.js was pretty clean, extract as much of the core logic to something more isolated

1. unless I'm missing something, it seems like a lot of state ends up on the rootScope (`$rootScope.modal = false;`, `$rootScope.timeoutModal`, `$rootScope.modalInstance`, `$rootScope.sessionTimeout`, `$rootScope.sessionWarning`, etc.). Unless these are being consumed by a directive or other template, don't do this - manage state elsewhere (like in a class or a closure)

**Comment by tbarsballe** [comment]
Composer has been moved to a private repo, and been updated with (a different implementation of) Gracefull Logout. Closing.

**Pull request closed without merging**